File: web-app/src/types/account.ts
/**
 * TypeScript types for account-related data structures
 */

export type AccountType =
  | 'checking'
  | 'savings'
  | 'brokerage'
  | 'investment'
  | 'credit_card'
  | 'loan'
  | '401k'
  | 'ira'
  | 'other';

export type BalanceType = 'statement' | 'calculated' | 'reconciled';

export interface Account {
  account_id: number;
  account_name: string;
  account_type: AccountType;
  account_subtype?: string;
  institution_name?: string;
  account_number_last4?: string;
  currency: string;
  is_active: boolean;
  has_cash_component: boolean;
  has_investment_component: boolean;
  opening_balance: number;
  opening_balance_date?: string;
  notes?: string;
  created_at: string;
  updated_at: string;
  // Current balance info (if available)
  current_balance?: number;
  current_balance_date?: string;
  cash_balance?: number;
  investment_value?: number;
  // Linked provider info
  link_type?: 'manual' | 'plaid' | 'akoya';
  linked_provider_id?: number;
}

export interface AccountCreate {
  account_name: string;
  account_type: AccountType;
  account_subtype?: string;
  institution_name?: string;
  account_number_last4?: string;
  currency?: string;
  opening_balance?: number;
  opening_balance_date?: string;
  notes?: string;
}

export interface AccountUpdate {
  account_name?: string;
  account_subtype?: string;
  institution_name?: string;
  account_number_last4?: string;
  is_active?: boolean;
  notes?: string;
}

export interface AccountSummary {
  account_id: number;
  account_name: string;
  account_type: string;
  institution?: string;
  current_balance: number;
  balance_date?: string;
  has_cash: boolean;
  has_investments: boolean;
  cash_balance?: number;
  investment_value?: number;
}

export interface Balance {
  balance_id: number;
  account_id: number;
  balance_date: string;
  total_balance: number;
  balance_type: BalanceType;
  cash_balance?: number;
  investment_value?: number;
  notes?: string;
  created_at: string;
  updated_at: string;
}

export interface BalanceCreate {
  balance_date: string;
  total_balance: number;
  balance_type?: BalanceType;
  cash_balance?: number;
  investment_value?: number;
  notes?: string;
}

export interface BalanceHistory {
  account_id: number;
  account_name: string;
  balances: Balance[];
  start_date?: string;
  end_date?: string;
}

export interface InvestmentHolding {
  holding_id: number;
  account_id: number;
  symbol: string;
  description?: string;
  quantity: number;
  cost_basis?: number;
  current_value?: number;
  as_of_date: string;
  asset_class?: string;
  sector?: string;
  created_at: string;
  updated_at: string;
  // Calculated fields
  gain_loss?: number;
  gain_loss_percent?: number;
}

export interface HoldingCreate {
  symbol: string;
  quantity: number;
  as_of_date: string;
  description?: string;
  cost_basis?: number;
  current_value?: number;
  asset_class?: string;
  sector?: string;
}

export interface PortfolioSummary {
  account_id: number;
  account_name: string;
  total_value: number;
  total_cost_basis?: number;
  total_gain_loss?: number;
  holdings: InvestmentHolding[];
  as_of_date: string;
}

export interface Reconciliation {
  reconciliation_id: number;
  account_id: number;
  statement_date: string;
  statement_balance: number;
  calculated_balance: number;
  statement_cash_balance?: number;
  calculated_cash_balance?: number;
  statement_investment_value?: number;
  calculated_investment_value?: number;
  discrepancy_amount?: number;
  is_reconciled: boolean;
  reconciled_at?: string;
  reconciled_by?: string;
  transactions_cleared_count: number;
  transactions_pending_count: number;
  notes?: string;
  created_at: string;
  updated_at: string;
}

export interface ReconciliationCreate {
  statement_date: string;
  statement_balance: number;
  statement_cash_balance?: number;
  statement_investment_value?: number;
  notes?: string;
}

export interface NetWorth {
  assets: number | string;
  liabilities: number | string;
  investments: number | string;
  // Support both snake_case (API) and camelCase (SDK) formats
  net_worth?: number | string;
  netWorth?: number | string;
  liquid_assets?: number | string;
  liquidAssets?: number | string;
  as_of_date?: string;
  asOfDate?: string;
  account_breakdown?: Record<string, number | string>;
  accountBreakdown?: Record<string, number | string>;
}

// Helper functions for account types
export const getAccountTypeLabel = (type: AccountType): string => {
  const labels: Record<AccountType, string> = {
    checking: 'Checking',
    savings: 'Savings',
    brokerage: 'Brokerage',
    investment: 'Investment',
    credit_card: 'Credit Card',
    loan: 'Loan',
    '401k': '401(k)',
    ira: 'IRA',
    other: 'Other',
  };
  return labels[type] || type;
};

export const getAccountTypeIcon = (type: AccountType): string => {
  const icons: Record<AccountType, string> = {
    checking: '🏦',
    savings: '💰',
    brokerage: '📈',
    investment: '📊',
    credit_card: '💳',
    loan: '🏠',
    '401k': '🏢',
    ira: '📋',
    other: '📁',
  };
  return icons[type] || '📁';
};

export const isAssetAccount = (type: AccountType): boolean => {
  return !['credit_card', 'loan'].includes(type);
};

export const isLiabilityAccount = (type: AccountType): boolean => {
  return ['credit_card', 'loan'].includes(type);
};

File: core-api/src/financial_analysis/api/schemas/account.py
"""
Pydantic schemas for account-related API endpoints.

These schemas define the request/response models for account management,
balance tracking, holdings, and reconciliation endpoints.
"""

from datetime import date, datetime
from decimal import Decimal
from typing import Optional, List, Literal
from pydantic import BaseModel, Field, ConfigDict


# ==================== Account Schemas ====================

class AccountBase(BaseModel):
    """Base account schema with common fields."""
    account_name: str = Field(..., min_length=1, max_length=100)
    account_type: Literal[
        'checking', 'savings', 'brokerage', 'investment',
        'credit_card', 'loan', '401k', 'ira', 'other'
    ]
    account_subtype: Optional[str] = Field(None, max_length=50)
    institution_name: Optional[str] = Field(None, max_length=100)
    account_number_last4: Optional[str] = Field(None, pattern=r'^\d{4}$')
    currency: str = Field(default='USD', pattern=r'^[A-Z]{3}$')
    notes: Optional[str] = None


class AccountCreate(AccountBase):
    """Schema for creating a new account."""
    opening_balance: Decimal = Field(default=Decimal('0'))
    opening_balance_date: Optional[date] = None
    entity_id: Optional[int] = Field(None, gt=0, description="Entity this account belongs to")


class AccountUpdate(BaseModel):
    """Schema for updating an account."""
    account_name: Optional[str] = Field(None, min_length=1, max_length=100)
    account_subtype: Optional[str] = Field(None, max_length=50)
    institution_name: Optional[str] = Field(None, max_length=100)
    account_number_last4: Optional[str] = Field(None, pattern=r'^\d{4}$')
    is_active: Optional[bool] = None
    entity_id: Optional[int] = Field(None, gt=0, description="Entity this account belongs to")
    notes: Optional[str] = None


class AccountResponse(AccountBase):
    """Schema for account responses."""
    model_config = ConfigDict(from_attributes=True)

    account_id: int
    is_active: bool
    has_cash_component: bool
    has_investment_component: bool
    opening_balance: Decimal
    opening_balance_date: Optional[date]
    created_at: datetime
    updated_at: datetime

    # Include current balance if available
    current_balance: Optional[Decimal] = None
    current_balance_date: Optional[date] = None
    cash_balance: Optional[Decimal] = None
    investment_value: Optional[Decimal] = None

    # Entity assignment
    entity_id: Optional[int] = None

    # Linked provider info
    link_type: str = 'manual'
    linked_provider_id: Optional[int] = None


class AccountSummary(BaseModel):
    """Schema for account summary with balance."""
    account_id: int
    account_name: str
    account_type: str
    institution: Optional[str]
    current_balance: Decimal
    balance_date: Optional[date]
    has_cash: bool
    has_investments: bool
    cash_balance: Optional[Decimal] = None
    investment_value: Optional[Decimal] = None


# ==================== Balance Schemas ====================

class BalanceCreate(BaseModel):
    """Schema for creating a balance snapshot."""
    balance_date: date
    total_balance: Decimal
    balance_type: Literal['statement', 'calculated', 'reconciled'] = 'statement'
    cash_balance: Optional[Decimal] = None
    investment_value: Optional[Decimal] = None
    notes: Optional[str] = None


class BalanceResponse(BaseModel):
    """Schema for balance responses."""
    model_config = ConfigDict(from_attributes=True)

    balance_id: int
    account_id: int
    balance_date: date
    total_balance: Decimal
    balance_type: str
    cash_balance: Optional[Decimal]
    investment_value: Optional[Decimal]
    notes: Optional[str]
    created_at: datetime
    updated_at: datetime


class BalanceHistory(BaseModel):
    """Schema for balance history response."""
    account_id: int
    account_name: str
    balances: List[BalanceResponse]
    start_date: Optional[date]
    end_date: Optional[date]


# ==================== Investment Holding Schemas ====================

class HoldingCreate(BaseModel):
    """Schema for creating/updating a holding."""
    symbol: str = Field(..., min_length=1, max_length=20)
    quantity: Decimal = Field(..., gt=0)
    as_of_date: date
    description: Optional[str] = Field(None, max_length=200)
    cost_basis: Optional[Decimal] = None
    current_value: Optional[Decimal] = None
    asset_class: Optional[str] = Field(None, max_length=50)
    sector: Optional[str] = Field(None, max_length=50)


class HoldingResponse(BaseModel):
    """Schema for holding responses."""
    model_config = ConfigDict(from_attributes=True)

    holding_id: int
    account_id: int
    symbol: str
    description: Optional[str]
    quantity: Decimal
    cost_basis: Optional[Decimal]
    current_value: Optional[Decimal]
    as_of_date: date
    asset_class: Optional[str]
    sector: Optional[str]
    created_at: datetime
    updated_at: datetime

    # Calculated fields
    gain_loss: Optional[Decimal] = None
    gain_loss_percent: Optional[float] = None


class PortfolioSummary(BaseModel):
    """Schema for portfolio summary."""
    account_id: int
    account_name: str
    total_value: Decimal
    total_cost_basis: Optional[Decimal]
    total_gain_loss: Optional[Decimal]
    holdings: List[HoldingResponse]
    as_of_date: date


# ==================== Reconciliation Schemas ====================

class ReconciliationCreate(BaseModel):
    """Schema for creating a reconciliation."""
    statement_date: date
    statement_balance: Decimal
    statement_cash_balance: Optional[Decimal] = None
    statement_investment_value: Optional[Decimal] = None
    notes: Optional[str] = None


class ReconciliationComplete(BaseModel):
    """Schema for completing a reconciliation."""
    reconciled_by: Optional[str] = Field(None, max_length=100)
    cleared_transaction_ids: Optional[List[int]] = None


class ReconciliationResponse(BaseModel):
    """Schema for reconciliation responses."""
    model_config = ConfigDict(from_attributes=True)

    reconciliation_id: int
    account_id: int
    statement_date: date
    statement_balance: Decimal
    calculated_balance: Decimal
    statement_cash_balance: Optional[Decimal]
    calculated_cash_balance: Optional[Decimal]
    statement_investment_value: Optional[Decimal]
    calculated_investment_value: Optional[Decimal]
    discrepancy_amount: Optional[Decimal]
    is_reconciled: bool
    reconciled_at: Optional[datetime]
    reconciled_by: Optional[str]
    transactions_cleared_count: int
    transactions_pending_count: int
    notes: Optional[str]
    created_at: datetime
    updated_at: datetime


# ==================== Net Worth & Analytics Schemas ====================

class NetWorthResponse(BaseModel):
    """Schema for net worth response."""
    assets: Decimal
    liabilities: Decimal
    net_worth: Decimal
    liquid_assets: Decimal
    investments: Decimal
    as_of_date: date

    # Breakdown by account type
    account_breakdown: Optional[dict] = None


class AccountFilterParams(BaseModel):
    """Schema for account filtering parameters."""
    is_active: Optional[bool] = None
    account_type: Optional[str] = None
    institution: Optional[str] = None
    has_cash: Optional[bool] = None
    has_investments: Optional[bool] = None


class TransactionClearRequest(BaseModel):
    """Schema for clearing transactions."""
    transaction_ids: List[int]
    cleared_date: Optional[date] = None


class CalculatedBalance(BaseModel):
    """Schema for calculated balance response."""
    account_id: int
    as_of_date: date
    total: Decimal
    cash: Optional[Decimal] = None
    investments: Optional[Decimal] = None
    based_on_transactions: int